File: 정종길/프로그래머스/석유 시추.py
from collections import deque, defaultdict

def solution(land):
    answer = defaultdict(int)

    r = len(land)
    c = len(land[0])
    q = deque()
    visited = [[0] * c for _ in range(r)]
    
    for i in range(r):
        for j in range(c):
            if not visited[i][j] and land[i][j]:
                visited[i][j] = 1
                q.append((i, j))
                column = set()
                cnt = 0
                while q:
                    x, y = q.popleft()
                    column.add(y)
                    cnt += 1
                    for dx, dy in (0, 1), (1, 0), (0, -1), (-1, 0):
                        nx, ny = x + dx, y + dy
                        if 0 <= nx < r and 0 <= ny < c:
                            if not visited[nx][ny] and land[nx][ny]:
                                visited[nx][ny] = 1
                                q.append((nx, ny))
                
                for col in column:
                    answer[col] += cnt

    return max(answer.values())

File: 정종길/프로그래머스/석유 시추.java
import java.util.*;

class Solution {

    static int n, m;
    static int[] oil;

    public int solution(int[][] land) {
        int answer = 0;

        n = land.length;
        m = land[0].length;
        oil = new int[m];

        boolean[][] visited = new boolean[n][m];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (land[i][j] == 1 && visited[i][j] == false) {
                    bfs(land, visited, i, j);
                }
            }
        }

        answer = Arrays.stream(oil).max().getAsInt();
        System.out.println(answer);
        return answer;
    }

    public void bfs(int[][] land, boolean[][] visited, int x, int y) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{x, y});
        visited[x][y] = true;


        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};


        int count = 1;

        Set<Integer> set = new HashSet<>();

        while (!q.isEmpty()) {
            int[] now = q.poll();
            set.add(now[1]);

            for (int i = 0; i < 4; i++) {
                int nx = now[0] + dx[i];
                int ny = now[1] + dy[i];

 
                if (!checkRange(nx, ny)) {
                    continue;
                }


                if (land[nx][ny] == 1 && visited[nx][ny] == false) {
                    q.add(new int[]{nx, ny});
                    visited[nx][ny] = true;
                    count += 1;
                }
            }
        }

        for (int index : set) {
            oil[index] += count;
        }
    }

    public boolean checkRange(int x, int y) {
        if (x < 0 || x >= n || y < 0 || y >= m) {
            return false;
        }
        return true;
    }
}